### DISCUSSION TITLE
Draw and Update are called at same frequency now?

### DISCUSSION THREAD

**Discussion by dazoe** [hajimehoshi/ebiten#1486] [Q&A] [answered]
While experimenting with setting lower TPS values i noticed some odd behavior.
when tps is set <=5 update is only called once at the start.
that was on v2.0.4
so I switched to using master, low TPS settings seem to work correctly now but then i noticed that Draw is called at the same freq as Update
git bisect found 4d545b9a was the reason for this change.
Was this intended? From what i understand from comparing with other game engines Update should be called at a set frequency for physics processing and things and Draw should run at target FPS (vsync) or as fast as possible if target FPS is not set, for smooth animations and such.

**Comment by hajimehoshi** [comment, marked as answer]
> Was this intended?

Yes. Assuming that no state is updated if Update is not called, Ebiten renders the previous offscreen without changing in this case.

**Reply by hajimehoshi** [reply]
On second thought, this is not a trivial behavior.

 * I assumed that Draw should not update the state, then the rendering result should be the same if Update is not called.
 * Users might assume that Draw is called based on FPS, and should not be affected by whether Update is called or not.

Hmm, let me think 🤔

**Reply by dazoe** [reply]
I still a noob when it comes to game dev but, many places I look say do your state update then render the updates. then go on to say run update at a fixed timestep and the most common is 60fps. then draw your state at vsync fps. But what i don't get is now we have computers, monitors, etc that run at 120, 144, and higher fps. so you end up drawing the same frame multiple times.
It seems to me that update should be called at a fixed time step for physics calculations, and draw should be used to handle everything else.
Unity for example has Update and FixedUpdate. Update is called per frame, the equivalent of ebiten's Draw function, and FixedUpdate is called on a set timestep, equivalent of ebiten's Update.
Godot has _process and _physics_process.

**Reply by hajimehoshi** [reply]
OK I committed a change to revert this optimization: 0b30ed595955b5f3d76808c135f739a0ec29944b

I understand users might want to use Draw with the time intervals of FPS, and Draw should not be skipped unexpectedly.

Thank you for the opinion!
